fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAutoCompleteResultCursor.java

```diff
@@ -0,0 +1,183 @@
+import com.apple.foundationdb.record.IndexEntry;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.RecordCursorContinuation;
+import com.apple.foundationdb.record.RecordCursorProto;
+import com.apple.foundationdb.record.RecordCursorResult;
+import com.apple.foundationdb.record.RecordCursorVisitor;
+import com.apple.foundationdb.record.ScanProperties;
+import com.apple.foundationdb.record.cursors.BaseCursor;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.provider.foundationdb.FDBStoreTimer;
+import com.apple.foundationdb.record.provider.foundationdb.IndexMaintainerState;
+import com.apple.foundationdb.tuple.Tuple;
+import com.google.protobuf.ByteString;
+import org.apache.lucene.search.suggest.Lookup;
+import org.apache.lucene.search.suggest.analyzing.AnalyzingInfixSuggester;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.IOUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+/**
+ * This class is a Record Cursor implementation for Lucene auto complete suggestion lookup.
+ * Because use cases of auto complete never need to get a big number of suggestions in one call, no scan with continuation support is needed.
+ * Suggestion result is populated as an {@link IndexEntry}, the key is in {@link IndexEntry#getKey()}, the field where it is indexed from and the value are in {@link IndexEntry#getValue()}.
+ */
+public class LuceneAutoCompleteResultCursor implements BaseCursor<IndexEntry> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneAutoCompleteResultCursor.class);
+    @Nonnull
+    private final Executor executor;
+    @Nonnull
+    private final IndexMaintainerState state;
+    @Nonnull
+    private final String query;
+    @Nullable
+    private final FDBStoreTimer timer;
+    private int limit;
+    @Nullable
+    private List<Lookup.LookupResult> lookupResults = null;
+    private int currentPosition;
+    @Nonnull
+    private final String searchTerm;
+    @Nullable
+    private final String fieldName;
+    @Nonnull
+    private final AnalyzingInfixSuggester suggester;
+    private final boolean highlight;
+
+    public LuceneAutoCompleteResultCursor(@Nonnull AnalyzingInfixSuggester suggester, @Nonnull String query,
+                                          @Nonnull Executor executor, @Nonnull ScanProperties scanProperties,
+                                          @Nonnull IndexMaintainerState state, boolean highlight) {
+        this.suggester = suggester;
+        this.query = query;
+        if (query.contains(":")) {
+            String[] subStrings = query.split(":", 2);
```

Comment:
should we check here if it is passing a more complex query?

Reply:
Although we follow the same syntax for the lucene function, I guess we don't have to support more complex query than a search term plus an optional field, for the auto-complete case.

Reply:
Would it be better to have two different query functions, rather than having anything we need to parse?

Isn't there an issue of completing over all fields multiple times and having a previous completion contain a colon in the middle? For example, suppose with subjects:
* **please** ⟶ Please read: agenda for 1/
* Please read: agenda for 1/**2** ⟶ Please read: agenda for 1/28

Reply:
TODO for later: lets make sure this second search term isn't a more complex query. for example a fuzzy search shouldn't be accepted but thats a future issue. For now this is great.

Reply:
@MMcM makes a good point. Honestly we don't need to accommodate any field matching here since we have the option to pass the list of fields we want to query through the luceneQueryComponent. I guess it would be easier to just get rid of the parsing altogether.